### DISCUSSION TITLE
The "Toolbox Drawer" idea

### DISCUSSION THREAD

**Discussion by poirazis** [Budibase/budibase#5630] [Ideas]
The Toolbox Drawer, would allow the users to do the following
- Save and Reuse code snippets ( JS / CSS )
- Save and Reuse Palettes and Colours
- Save and Reuse groups of components ( That nice multi column container you designed , save it, clone it again )
- Save and Reuse screes ( You design you nice modal details popup, save it, clone it or inherit it )
- Save and Reuse Builder Scripts, this is the interesting part. Scripts would be sequences of actions defined within bb Builder to automate tasks. Autogenerated CRUD screens is the prime example. But why not let people build their own ? 

Its really like a large, categorised, persistable clipboard.

Just putting it out there

**Comment by melohagan** [comment]
Some great suggestions here, thanks for sharing!

The JS code reuse is a nice one, and has been suggested in the past: https://github.com/Budibase/budibase/issues/3571
but as per this poll it's likely the team will work on custom components first: https://github.com/Budibase/budibase/discussions/5518
I think that'd cover your third bullet point. 

You can duplicate screens and duplicate containers (including their children) which might be useful too.

Being able to apply themes would be nice. It's explored a bit here: https://github.com/Budibase/budibase/issues/3370

And for builder scripts. That is an interesting one; I wonder what common repetitive tasks users find themselves doing. Is there anything in particular you would use these builder scripts for?

**Reply by poirazis** [reply]
Hey @melohagan 

well , the very first thing i would do would be to duplicate the crud autogeneration workflow and customise to my design style. to create the screens the way i would want them. and i would have different workflows saved.

I would be able to script "mass updates" on various components, i.e change the gradient on outer container in any group of screens.

you could also probably script some conversions / operations you do on elements. like convert an element to another type. or split a large field group in to a container -> section (2 column) -> Field Group 1 & Filed Group 2.

I mean having a scriptable environment opens up all sorts of possibilities i guess.

i am more brainstorming than specifically suggesting i guess.
